### ISSUE TITLE
[Metrics] Bus Factor

### ISSUE THREAD

**Issue by xiaoya-Esther** [X-lab2017/open-digger#594]
**Description**
- How high is the risk to a project should the most active people leave?
- The Bus Factor is the smallest number of people that make 50% of contributions.

**Objective**
- Identify how widely the work in a project is distributed across contributors. 
- Identify the key people in a project that are doing the majority of the work.

**Implementation**
Adding up each contributor's contributions sorted in decreasing order until we reach the threshold(50% of all contributions).

![image](https://user-images.githubusercontent.com/56542972/120434390-037aba80-c3af-11eb-9a92-cc6a9c3f5837.png)
**The Bus Factor for the CHAOSS Project in 2020, when considering only git commits, was 5.**

**Resource**
CHAOSS: Evolution -> Community Growth
https://chaoss.community/metric-bus-factor/

**Comment by frank-zsy** [comment]
CHAOSS's default implementation of bus factor uses commit count as the number of contributions. But in my observation, commit is not a very good metrics for this factor.

The reasons is commit from original repo will exist in a forked repo. Like all the Linux kernel based project, all commit history is asked to preserve so we can not easily distinguish the contribution of new contributions to a certain downstream repo from the main branch of kernel.

So I think merged change request is a better metrics for bus factor and we may also need to filter some accounts like robots in this one.

**Comment by xiaoya-yaya** [comment]
I like the idea. I think when calculating the bus factor, we need to think clearly about what means "important contributions that keep the project running", because we are trying to find out who really makes it run, and without them there is a high possibility of project downtime. So what I'm not sure about is:
- The percentage of projects whose contributors with permission are used of direct committing (how many direct commits might be left out)
- The percentage of forked projects who care about their bus factors (that it's a hard fork instead of just a patch that intends to be merged upstream)

It's a trade-off.

**Comment by frank-zsy** [comment]
For the first situation, many open source projects especially well collaborated projects, they need to use change request to merge commits which takes advantage of the permission control system of the platform. Directly commit is not quite friendly to collaboration. But indeed using change request has problems with the project always using direct commit like kernel. So maybe we can implement as using change request by default and can use commit information by config as well.

I don't quite understand the second one. For a hard fork, the following maintain work is just about the repo itself. So if we count the contributions of the forked project, the result makes sense in my view because for a hard fork, the bus factor should be only related to the repo rather than the upstream repo.

**Comment by xiaoya-yaya** [comment]
I agree that the user can choose to count `commit` or `change request` by config.
We can start with implementing change requests by default and commit by config.
